### PR TITLE
Fix and simplify datastore query result pagination

### PR DESCRIPTION
**Pull request by jey**
The previous assertion `isLast != moreBatches` failed spuriously.

I've run the new `bulk_query` test against Cloud Datastore, but not the Datastore Emulator.

### CODE REVIEW

**Comment by jey** [comment]
The unit test is a bit flaky, the `waitUntilEntitiesReady` and `waitUntilEntitiesGone` calls are timing out sometimes.

File: lib/src/datastore_impl.dart [hunk 1]
```diff
@@ -541,27 +529,16 @@ class QueryPageImpl implements Page<datastore.Entity> {
...
+      // Is this the last page of results for this query?
+      bool isLast = response.batch.moreResults == 'NO_MORE_RESULTS' ||
           response.batch.moreResults == 'MORE_RESULTS_AFTER_LIMIT';
 
-      bool gotAll = limit != null && remainingEntities == 0;
-      bool noMore = response.batch.moreResults == 'NO_MORE_RESULTS';
-      bool isLast = gotAll || noMore;
+      // Either this is the last page, xor the state should be 'NOT_FINISHED'.
+      // Other states like 'MORE_RESULTS_AFTER_CURSOR' are unexpected.
+      assert(isLast != (response.batch.moreResults == 'NOT_FINISHED'));
```
[the remark this answers is not in the document]
Comment (jey) [reply lib/src/datastore_impl.dart]: If all is working as expected, then this assertion will always be true. However, there are other status codes that could be returned by the API like `MORE_RESULTS_AFTER_LIMIT`. In that case it becomes `assert(false != false)`, which fails.

An alternative (and possibly clearer) way of writing this check would be:
```suggestion
      assert(isLast || response.batch.moreResults == 'NOT_FINISHED');
```

I'll go ahead and make that change :)

[the remark this answers is not in the document]
Comment (jonasfj) [reply lib/src/datastore_impl.dart]: Ah, yes, that makes more sense..

File: lib/src/datastore_impl.dart [hunk 2]
```diff
@@ -541,27 +529,16 @@ class QueryPageImpl implements Page<datastore.Entity> {
...
 
-      // As a sanity check, we assert that `moreBatches XOR isLast`.
-      assert(isLast != moreBatches);
-
-      // FIXME: TODO: Big hack!
-      // It looks like Apiary/Atlas is currently broken.
-      if (moreBatches && returnedEntities.isEmpty) {
-        print('Warning: Api to Google Cloud Datastore returned bogus response. '
```
[the remark this answers is not in the document]
Comment (jey) [reply lib/src/datastore_impl.dart]: I have not encountered this warning. Also, I don't think the `moreBatches` logic is actually correct. The pre-existing assertion `assert(isLast != moreBatches)` failed when doing a regular paginated query. (This is demonstrated by the new `bulk_query` test.)

File: lib/src/datastore_impl.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (jey) [reply lib/src/datastore_impl.dart]: Fixed

**Pull request closed without merging**